File: examples/excel-replace.ts
/**
 * 例子：打开 xlsx，把每个含 `{{client}}` 的 Cell 文本替换为指定值，写回新文件。
 *
 * 仅替换 `dataType="str"` / `"inlineStr"` / 默认无 dataType（即不走 sharedString）的
 * cellValue.text；走 sharedString 的 cell 需要更复杂的 intern + 索引更新——本例
 * 不覆盖该路径。
 *
 * 跑法：
 *   bun run examples/excel-replace.ts <input.xlsx> <output.xlsx> [<client-name>]
 */

import { Cell, CellValue, SpreadsheetDocument } from "../src/excel/index.js";

async function main(): Promise<void> {
  const [inputPath, outputPath, client = "Acme Corp"] = process.argv.slice(2);
  if (inputPath === undefined || outputPath === undefined) {
    process.stderr.write("usage: excel-replace <input.xlsx> <output.xlsx> [<client-name>]\n");
    process.exit(2);
  }

  const doc = await SpreadsheetDocument.openAsync(inputPath);
  const wp = doc.workbookPart;
  if (wp === undefined) {
    process.stderr.write("error: input is not a valid xlsx (no workbook part)\n");
    process.exit(1);
  }

  let replaced = 0;
  for (const wsp of wp.worksheetParts) {
    for (const c of wsp.worksheet.descendants(Cell)) {
      const v = c.firstChild(CellValue);
      if (v?.text?.includes("{{client}}") === true) {
        v.text = v.text.replaceAll("{{client}}", client);
        replaced += 1;
      }
    }
  }

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Replaced ${replaced} occurrence(s); wrote ${outputPath}\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelReplace.cs
// Replica of examples/excel-replace.ts
// Creates an xlsx template with "{{client}}" placeholder, opens it, replaces with "Acme Corp".

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelReplace
{
    private const string Client = "Acme Corp";

    public static void Run(string outputPath)
    {
        var tempPath = Path.GetTempFileName() + ".xlsx";
        try
        {
            BuildTemplate(tempPath);
            ApplyReplace(tempPath, outputPath);
        }
        finally
        {
            if (File.Exists(tempPath)) File.Delete(tempPath);
        }
    }

    private static void BuildTemplate(string path)
    {
        using var doc = SpreadsheetDocument.Create(path,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();
        var sheetData = new SheetData();
        wsPart.Worksheet = new Worksheet(sheetData);

        wbPart.Workbook.AppendChild(new Sheets()).Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
        });

        // Two rows with {{client}} placeholder
        var r1 = new Row { RowIndex = 1 };
        r1.Append(MakeStrCell("A1", "Dear {{client}},"));
        sheetData.Append(r1);

        var r2 = new Row { RowIndex = 2 };
        r2.Append(MakeStrCell("A2", "Thank you, {{client}}."));
        sheetData.Append(r2);

        wbPart.Workbook.Save();
    }

    private static void ApplyReplace(string inputPath, string outputPath)
    {
        File.Copy(inputPath, outputPath, overwrite: true);

        using var doc = SpreadsheetDocument.Open(outputPath, true);
        var wb = doc.WorkbookPart;
        if (wb == null) return;

        foreach (var wsp in wb.WorksheetParts)
        {
            foreach (var cell in wsp.Worksheet.Descendants<Cell>())
            {
                var v = cell.CellValue;
                if (v?.Text?.Contains("{{client}}") == true)
                    v.Text = v.Text.Replace("{{client}}", Client);
            }
        }

        wb.Workbook.Save();
    }

    private static Cell MakeStrCell(string cellRef, string value) =>
        new Cell
        {
            CellReference = cellRef,
            DataType = CellValues.String,
            CellValue = new CellValue(value)
        };
}
